File: ruby/triangle/triangle.rb
class Triangle

    def initialize sides
        @sides      = sides
        @uniq_sides = sides.uniq
    end

    def equilateral?
        three_sides_equal? && ! all_sides_are_zero?
    end

    def isosceles?
        valid_triangle? && (two_sides_equal? || equilateral? )
    end

    def scalene?
        valid_triangle? && all_sides_different?
    end
    private

    attr_reader :sides, :uniq_sides

    def valid_triangle?
        ( sides[0] + sides[1] >= sides[2] ) &&
        ( sides[0] + sides[2] >= sides[1] ) &&
        ( sides[1] + sides[2] >= sides[0] )   
    end

    def three_sides_equal?
        uniq_sides.count == 1
    end

    def two_sides_equal?
        uniq_sides.count == 2
    end

    def all_sides_different?
        uniq_sides.count == 3
    end

    def all_sides_are_zero?
        uniq_sides.count == 1 && uniq_sides.first == 0
    end
end

File: python/triangle/triangle.py
"""
    Triangle validations module
"""

def valid_triangle(f):
    """
    Decorator to verify that provided sides can form a triangle.
    """
    def inner(sides):
        (a,b,c) = sides

        if a == b == c == 0.0:
            return False

        return (a + b >= c) and (b + c >= a) and (a + c >= b) and f(sides)

    return inner

@valid_triangle
def equilateral(sides):
    """
    Returns True if the triangle is equilateral, False otherwise.
    """

    (a,b,c) = sides
    return a == b == c

@valid_triangle
def isosceles(sides):
    """
    Returns True if the triangle is isosceles, False otherwise.
    """

    (a,b,c) = sides
    return a == b or a == c or b == c
    # return len(set(sides)) <= 2

@valid_triangle
def scalene(sides):
    """
    Returns True if the triangle is scalene, False otherwise.
    """

    (a,b,c) = sides
    return a != b and a != c and b != c
